File: internal/setup/setup_test.go
package setup

import (
	"os"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestSetupDocs(t *testing.T) {
	tmpDir := t.TempDir()
	originalDir, err := os.Getwd()
	require.NoError(t, err)
	defer os.Chdir(originalDir)

	err = os.Chdir(tmpDir)
	require.NoError(t, err)

	err = SetupDocs("test-project")
	require.NoError(t, err)

	// Check that directories were created
	dirs := []string{
		"docs",
		"docs/adr",
		"docs/api",
		"docs/guides",
		"docs/prompts",
		"docs/development",
	}

	for _, dir := range dirs {
		assert.DirExists(t, dir, "Directory %s should exist", dir)
	}

	// Check that key files were created
	files := []string{
		"mkdocs.yml",
		"docs/index.md",
		"docs/adr/index.md",
		"docs/adr/001-use-mkdocs.md",
		"docs/adr/002-perfect-commits.md",
		"docs/adr/003-track-ai-prompts.md",
		"docs/development/setup.md",
		"docs/development/contributing.md",
		"docs/development/testing.md",
		"docs/prompts/index.md",
		"docs/guides/index.md",
		"docs/requirements.txt",
	}

	for _, file := range files {
		assert.FileExists(t, file, "File %s should exist", file)
	}

	// Verify mkdocs.yml contains project name
	content, err := os.ReadFile("mkdocs.yml")
	require.NoError(t, err)
	assert.Contains(t, string(content), "test-project")
}

func TestSetupCommitTemplates(t *testing.T) {
	tmpDir := t.TempDir()
	originalDir, err := os.Getwd()
	require.NoError(t, err)
	defer os.Chdir(originalDir)

	err = os.Chdir(tmpDir)
	require.NoError(t, err)

	// Initialize git
	err = InitGit()
	require.NoError(t, err)

	err = SetupCommitTemplates()
	require.NoError(t, err)

	// Check that .gitmessage was created
	assert.FileExists(t, ".gitmessage")

	content, err := os.ReadFile(".gitmessage")
	require.NoError(t, err)
	assert.Contains(t, string(content), "Perfect Commit Structure")
	assert.Contains(t, string(content), "Implementation:")
	assert.Contains(t, string(content), "Tests:")
	assert.Contains(t, string(content), "Documentation:")

	// Check that hooks were created
	hooks := []string{
		".git/hooks/commit-msg",
		".git/hooks/prepare-commit-msg",
		".git/hooks/pre-commit",
	}

	for _, hook := range hooks {
		assert.FileExists(t, hook, "Hook %s should exist", hook)

		// Check that hooks are executable
		info, err := os.Stat(hook)
		require.NoError(t, err)
		assert.NotEqual(t, 0, info.Mode()&0111, "Hook %s should be executable", hook)
	}
}

func TestSetupGithubActions(t *testing.T) {
	tmpDir := t.TempDir()
	originalDir, err := os.Getwd()
	require.NoError(t, err)
	defer os.Chdir(originalDir)

	err = os.Chdir(tmpDir)
	require.NoError(t, err)

	err = SetupGithubActions("test-project")
	require.NoError(t, err)

	// Check that workflows were created
	workflows := []string{
		".github/workflows/ci.yml",
		".github/workflows/docs.yml",
		".github/workflows/release.yml",
	}

	for _, workflow := range workflows {
		assert.FileExists(t, workflow, "Workflow %s should exist", workflow)
	}

	// Check that templates were created
	assert.FileExists(t, ".github/PULL_REQUEST_TEMPLATE.md")
	assert.FileExists(t, ".github/ISSUE_TEMPLATE/bug_report.md")
	assert.FileExists(t, ".github/ISSUE_TEMPLATE/feature_request.md")

	// Verify CI workflow content
	ciContent, err := os.ReadFile(".github/workflows/ci.yml")
	require.NoError(t, err)
	assert.Contains(t, string(ciContent), "go test")
	assert.Contains(t, string(ciContent), "golangci-lint")
}

func TestSetupProjectStructure(t *testing.T) {
	tmpDir := t.TempDir()
	originalDir, err := os.Getwd()
	require.NoError(t, err)
	defer os.Chdir(originalDir)

	err = os.Chdir(tmpDir)
	require.NoError(t, err)

	err = SetupProjectStructure("test-project")
	require.NoError(t, err)

	// Check that directories were created
	dirs := []string{
		"cmd",
		"internal",
		"pkg",
		"test",
		"scripts",
	}

	for _, dir := range dirs {
		assert.DirExists(t, dir, "Directory %s should exist", dir)
	}

	// Check that files were created
	files := []string{
		".gitignore",
		"Makefile",
		".golangci.yml",
		"CONTRIBUTING.md",
	}

	for _, file := range files {
		assert.FileExists(t, file, "File %s should exist", file)
	}

	// Verify .gitignore content
	gitignoreContent, err := os.ReadFile(".gitignore")
	require.NoError(t, err)
	assert.Contains(t, string(gitignoreContent), "*.exe")
	assert.Contains(t, string(gitignoreContent), ".DS_Store")
}

func TestCreateReadme(t *testing.T) {
	tmpDir := t.TempDir()
	originalDir, err := os.Getwd()
	require.NoError(t, err)
	defer os.Chdir(originalDir)

	err = os.Chdir(tmpDir)
	require.NoError(t, err)

	err = CreateReadme("test-project")
	require.NoError(t, err)

	assert.FileExists(t, "README.md")

	content, err := os.ReadFile("README.md")
	require.NoError(t, err)
	assert.Contains(t, string(content), "test-project")
	assert.Contains(t, string(content), "Perfect Commits")
	assert.Contains(t, string(content), "Simon Willison")
}

func TestInitGit(t *testing.T) {
	tmpDir := t.TempDir()
	originalDir, err := os.Getwd()
	require.NoError(t, err)
	defer os.Chdir(originalDir)

	err = os.Chdir(tmpDir)
	require.NoError(t, err)

	err = InitGit()
	require.NoError(t, err)

	// Check that .git directory was created
	assert.DirExists(t, ".git")

	// Running again should not error
	err = InitGit()
	require.NoError(t, err)
}

File: tests/setup.test.ts
import { describe, test, expect, beforeEach, afterEach } from 'bun:test';
import { mkdir, rm, readFile, access } from 'node:fs/promises';
import { setupDocs } from '../src/setup/docs';
import { setupCommitTemplates } from '../src/setup/commits';
import { setupGithubActions } from '../src/setup/github';
import { setupProjectStructure } from '../src/setup/project';
import { createReadme } from '../src/setup/readme';
import { initGit } from '../src/setup/git';

const TEST_DIR = '/tmp/ai-test-' + Date.now();

beforeEach(async () => {
  await mkdir(TEST_DIR, { recursive: true });
  process.chdir(TEST_DIR);
});

afterEach(async () => {
  try {
    await rm(TEST_DIR, { recursive: true, force: true });
  } catch {
    // Ignore cleanup errors
  }
});

describe('setupDocs', () => {
  test('creates documentation structure', async () => {
    await setupDocs('test-project');

    // Check directories exist
    await access('docs');
    await access('docs/.vitepress');
    await access('docs/adr');
    await access('docs/development');
    await access('docs/guides');
    await access('docs/prompts');

    // Check VitePress config exists
    const vitepressConfig = await readFile('docs/.vitepress/config.ts', 'utf-8');
    expect(vitepressConfig).toContain('test-project');

    const indexContent = await readFile('docs/index.md', 'utf-8');
    expect(indexContent).toContain('test-project');

    // Check ADRs
    await access('docs/adr/001-use-vitepress.md');
    await access('docs/adr/004-use-bun-typescript.md');
  });
});

describe('setupCommitTemplates', () => {
  test('creates commit templates and hooks', async () => {
    await initGit();
    await setupCommitTemplates();

    // Check .gitmessage exists
    const gitmessage = await readFile('.gitmessage', 'utf-8');
    expect(gitmessage).toContain('Perfect Commit Structure');

    // Check hooks exist
    await access('.git/hooks/commit-msg');
    await access('.git/hooks/prepare-commit-msg');
    await access('.git/hooks/pre-commit');
  });
});

describe('setupGithubActions', () => {
  test('creates GitHub workflows', async () => {
    await setupGithubActions('test-project');

    // Check workflows
    await access('.github/workflows/ci.yml');
    await access('.github/workflows/docs.yml');
    await access('.github/workflows/release.yml');

    const ciContent = await readFile('.github/workflows/ci.yml', 'utf-8');
    expect(ciContent).toContain('bun');

    // Check templates
    await access('.github/PULL_REQUEST_TEMPLATE.md');
    await access('.github/ISSUE_TEMPLATE/bug_report.md');
  });
});

describe('setupProjectStructure', () => {
  test('creates project structure', async () => {
    await setupProjectStructure('test-project');

    // Check directories
    await access('src');
    await access('tests');

    // Check files
    await access('.gitignore');
    await access('Makefile');

    const gitignore = await readFile('.gitignore', 'utf-8');
    expect(gitignore).toContain('node_modules');
  });
});

describe('createReadme', () => {
  test('creates README with project name', async () => {
    await createReadme('test-project');

    const readme = await readFile('README.md', 'utf-8');
    expect(readme).toContain('test-project');
    expect(readme).toContain('Perfect Commit');
    expect(readme).toContain('OpenTUI');
  });
});
